app/src/main/java/org/fedorahosted/freeotp/widget/OtpListWidgetViewsFactory.java

```diff
@@ -0,0 +1,110 @@
+package org.fedorahosted.freeotp.widget;
+
+import android.content.Context;
+import android.content.Intent;
+import android.graphics.Bitmap;
+import android.widget.RemoteViews;
+import android.widget.RemoteViewsService;
+import com.squareup.picasso.Picasso;
+import org.fedorahosted.freeotp.R;
+import org.fedorahosted.freeotp.Token;
+import org.fedorahosted.freeotp.TokenPersistence;
+import org.fedorahosted.freeotp.TokenPlaceholderGenerator;
+
+import java.io.IOException;
+
+/**
+ * Created by root on 13/04/17.
+ */
+public class OtpListWidgetViewsFactory implements RemoteViewsService.RemoteViewsFactory {
+
+    private final Context context;
+    private final TokenPersistence persistence;
+    private final int widgetId;
+
+    OtpListWidgetViewsFactory(final Context context, final int widgetId) {
+        this.context = context;
+        this.widgetId = widgetId;
+        persistence = new TokenPersistence(context);
+    }
+
+    @Override
+    public int getCount() {
+        return persistence.length();
+    }
+
+    @Override
+    public RemoteViews getViewAt(int position) {
+        final Token token = persistence.get(position);
+        final String tokenId = token.getID();
+        final RemoteViews row = new RemoteViews(context.getPackageName(), R.layout.widget_row);
+
+        try {
+            Bitmap b = Picasso.with(context).load(token.getImage()).get();
```

Comment:
This not how to use Picasso. Calling `.get()` will block the thread instead of taking advantage of assynchronous call of Picasso.

You may:
- use `.error(R.drawable.logo)` to handle the error (don't use try-catch when Picassso does it for you)
- use `.into(row, R.id.widget_image, ...)` to put the loaded image in the imageview